### PR TITLE
Declarative Testing

### PR DESCRIPTION
**Pull request by Jomik**

### CODE REVIEW

File: src/future.ts [hunk 1]
```diff
@@ -8,6 +8,15 @@ import { Stream } from "./stream";
 
 export type MapFutureTuple<A> = { [K in keyof A]: Future<A[K]> };
 
+export type SemanticFuture<A> =
+  | { time: number; value: A }
+  | { time: 9007199254740991; value: never };
+
+const neverOccuringSemanticFuture = {
+  time: 9007199254740991,
```
Comment (paldepind) [inline src/future.ts]: What about using `Infinity` instead of `9007199254740991`?

Comment (Jomik) [reply src/future.ts]: We could use `Infinity` in `neverOccuringSemanticFuture`, but `Infinity` is not a type, that is why we decided on this super large number which is the same as `Number.MAX_SAFE_INTEGER`.

https://github.com/Microsoft/TypeScript/issues/15135
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/MAX_SAFE_INTEGER

File: src/future.ts [hunk 2]
```diff
@@ -1,13 +1,24 @@
...
-import { Stream } from "./stream";
+import { Stream, Occurrence } from "./stream";
 
 export type MapFutureTuple<A> = { [K in keyof A]: Future<A[K]> };
 
+export function doesOccur<A>(
+  future: SemanticFuture<A>
+): future is Occurrence<A> {
+  return future.time !== "infinity";
+}
```
Comment (paldepind) [inline src/future.ts]: Missing linebreak

File: src/future.ts [hunk 3]
```diff
@@ -1,13 +1,24 @@
...
+export const neverOccurringFuture = {
+  time: "infinity" as "infinity",
+  value: undefined as undefined
+};
```
Comment (paldepind) [inline src/future.ts]: Missing linebreak.

File: src/now.ts [hunk 4]
```diff
@@ -31,9 +31,9 @@ export abstract class Now<A> implements Monad<A> {
   }
   static multi: boolean = false;
   multi: boolean = false;
-  test(t: Time): A {
-    throw new Error("The Now computation does not support testing yet");
-  }
+
+  abstract test(mocks: any[], t: Time): { value: A; mocks: any[] };
```
Comment (paldepind) [inline src/now.ts]: Two extra line breaks.

File: src/now.ts [hunk 5]
```diff
@@ -89,6 +90,10 @@ class PerformNow<A> extends Now<A> {
   run(): A {
     return this.cb();
   }
+
```
Comment (paldepind) [inline src/now.ts]: Extra linebreak.

File: src/now.ts [hunk 6]
```diff
@@ -118,6 +123,13 @@ class PerformMapNow<A, B> extends Now<Stream<B> | Future<B>> {
       ? mapCbStream((value, done) => done(this.cb(value)), this.s)
       : mapCbFuture((value, done) => done(this.cb(value)), this.s);
   }
+
```
Comment (paldepind) [inline src/now.ts]: Extra linebreak.

File: src/now.ts [hunk 7]
```diff
@@ -165,6 +178,19 @@ class PerformIOStreamLatest<A> extends ActiveStream<A>
...
 }
 
+class PerformStreamNowLatest<A> extends Now<Stream<A>> {
+  constructor(private s: Stream<IO<A>>) {
+    super();
+  }
+  run(): Stream<A> {
+    return new PerformIOStreamLatest(this.s);
+  }
+
```
Comment (paldepind) [inline src/now.ts]: Extra linebreak.

File: src/now.ts [hunk 8]
```diff
@@ -199,8 +225,34 @@ class PerformIOStreamOrdered<A> extends ActiveStream<A> {
...
 }
 
+class PerformStreamNowOrdered<A> extends Now<Stream<A>> {
+  constructor(private s: Stream<IO<A>>) {
+    super();
+  }
+  run(): Stream<A> {
+    return new PerformIOStreamOrdered(this.s);
+  }
+
```
Comment (paldepind) [inline src/now.ts]: Extra linebreak.

File: src/now.ts [hunk 9]
```diff
@@ -199,8 +225,34 @@ class PerformIOStreamOrdered<A> extends ActiveStream<A> {
...
+}
+
+class PlanNow<A> extends Now<Future<A>> {
+  constructor(private future: Future<Now<A>>) {
+    super();
+  }
+  run(time: Time): Future<A> {
+    return this.future.map((n) => n.run(time));
+  }
+
```
Comment (paldepind) [inline src/now.ts]: Extra linebreak.

File: src/test.ts [hunk 10]
```diff
@@ -36,3 +37,23 @@ export function testStreamFromObject<A>(object: {
...
+  constructor(private semanticFuture: SemanticFuture<A>) {
+    super();
+  }
+  pushS(t: number, val: A): void {
+    throw new Error("You cannot push to a TestFuture");
+  }
+  semantic(): SemanticFuture<A> {
+    return this.semanticFuture;
+  }
+
```
Comment (paldepind) [inline src/test.ts]: Extra linebreak.

**Pull request merged**